src/zarr/storage/_obstore.py

```diff
@@ -0,0 +1,419 @@
+from __future__ import annotations
+
+import asyncio
+import contextlib
+import pickle
+from collections import defaultdict
+from collections.abc import Iterable
+from typing import TYPE_CHECKING, Any, TypedDict
+
+from zarr.abc.store import (
+    ByteRequest,
+    OffsetByteRequest,
+    RangeByteRequest,
+    Store,
+    SuffixByteRequest,
+)
+from zarr.core.buffer import Buffer
+from zarr.core.buffer.core import BufferPrototype
+
+if TYPE_CHECKING:
+    from collections.abc import AsyncGenerator, Coroutine, Iterable
+    from typing import Any
+
+    from obstore import ListResult, ListStream, ObjectMeta, OffsetRange, SuffixRange
+    from obstore.store import ObjectStore as _UpstreamObjectStore
+
+    from zarr.core.buffer import BufferPrototype
+    from zarr.core.common import BytesLike
+
+__all__ = ["ObjectStore"]
+
+_ALLOWED_EXCEPTIONS: tuple[type[Exception], ...] = (
+    FileNotFoundError,
+    IsADirectoryError,
+    NotADirectoryError,
+)
+
+
+class ObjectStore(Store):
+    """A Zarr store that uses obstore for fast read/write from AWS, GCP, Azure.
+
+    Parameters
+    ----------
+    store : obstore.store.ObjectStore
+        An obstore store instance that is set up with the proper credentials.
+    read_only : bool
+        Whether to open the store in read-only mode.
+
+    Warnings
+    --------
+    ObjectStore is experimental and subject to API changes without notice. Please
+    raise an issue with any comments/concerns about the store.
+    """
+
+    store: _UpstreamObjectStore
+    """The underlying obstore instance."""
+
+    def __eq__(self, value: object) -> bool:
+        if not isinstance(value, ObjectStore):
+            return False
+
+        if not self.read_only == value.read_only:
+            return False
+
+        return self.store == value.store
+
+    def __init__(self, store: _UpstreamObjectStore, *, read_only: bool = False) -> None:
+        import obstore as obs
+
+        if not isinstance(
```

Comment:
What's the rationale for this check?

A potential downside is that obstore grows a new store implementation in the future, and we'll need to add it to the list allowed here.

The upside of this check is maybe catching an issue at `__init__` time rather than when you call a method (and it presumably fails. Unless it's duck typed appropriately).

If you want to keep this, then does obstore have a base class we could check against?

Reply:
There's a test for this below that would need to be updated.